gpu-switch.sh

```diff
@@ -0,0 +1,311 @@
+            fi
+        done
+        log "Updating Docker container with VFIO devices"
+        docker update $devices_args "$container_name" || \
+            error_exit "Failed to update Docker container devices"
+    else
+        log "Removing VFIO devices from Docker container"
+        for gpu in "${!GPU_PAIRS[@]}"; do
+            local audio="${GPU_PAIRS[$gpu]}"
+            docker update --device-rm "vfio-pci,host=$gpu" "$container_name" 2>/dev/null || \
+                log "Warning: Failed to remove GPU device from container"
+            if [ ! -z "$audio" ]; then
+                docker update --device-rm "vfio-pci,host=$audio" "$container_name" 2>/dev/null || \
+                    log "Warning: Failed to remove audio device from container"
+            fi
+        done
+    fi
+}
+
+verify_gpu_bound() {
+    local target_driver="$1"
+    log "Verifying device bindings to $target_driver"
+    
+    for gpu in "${!GPU_PAIRS[@]}"; do
+        local audio="${GPU_PAIRS[$gpu]}"
+        local current_gpu_driver=$(readlink "/sys/bus/pci/devices/$gpu/driver" 2>/dev/null | awk -F '/' '{print $NF}')
+        
+        if [[ "$target_driver" == "vfio-pci" ]]; then
+            if [[ "$current_gpu_driver" != "vfio-pci" ]]; then
+                error_exit "GPU $gpu failed to bind to VFIO-PCI (current: $current_gpu_driver)"
+            fi
+            
+            if [ ! -z "$audio" ]; then
+                local current_audio_driver=$(readlink "/sys/bus/pci/devices/$audio/driver" 2>/dev/null | awk -F '/' '{print $NF}')
+                if [[ "$current_audio_driver" != "vfio-pci" ]]; then
+                    error_exit "Audio device $audio failed to bind to VFIO-PCI (current: $current_audio_driver)"
+                fi
+            fi
+        elif [[ "$target_driver" == "nvidia" ]]; then
+            if [[ "$current_gpu_driver" != "nvidia" ]]; then
+                error_exit "GPU $gpu failed to bind to NVIDIA (current: $current_gpu_driver)"
+            fi
+            
+            if [ ! -z "$audio" ]; then
+                local current_audio_driver=$(readlink "/sys/bus/pci/devices/$audio/driver" 2>/dev/null | awk -F '/' '{print $NF}')
+                if [[ "$current_audio_driver" != "snd_hda_intel" ]]; then
+                    log "Warning: Audio device $audio not bound to snd_hda_intel (current: $current_audio_driver)"
+                fi
+            fi
+        fi
+    done
+    log "All devices successfully bound to $target_driver"
+}
+
+# Main script
+if [ "$#" -ne 2 ]; then
+    error_exit "Usage: $0 [start|stop] CONTAINER_NAME"
+fi
+
+ACTION="$1"
+CONTAINER_NAME="$2"
+
+# Verify container exists
+if ! docker ps -a --format '{{.Names}}' | grep -q "^${CONTAINER_NAME}$"; then
+    error_exit "Container $CONTAINER_NAME does not exist"
+fi
+
+# Check if GPU management is needed
+if ! check_gpu_needed; then
+    log "Continuing without GPU management"
+    exit 0
+fi
```

Comment:
Instead of listing all containers, you can directly check the existence of the container using docker container inspect, which is more efficient since it only checks the specified container without scanning the entire list. Here’s how to replace that line:

if ! docker container inspect "$CONTAINER_NAME" > /dev/null 2>&1; then
    error_exit "Container $CONTAINER_NAME does not exist"
fi

The docker container inspect command returns an error if the container does not exist, so you can use it to directly verify the container’s existence without listing all containers.

Reply:
ill take a look at implementing this